### ISSUE TITLE
Problem with regexp added when saving a page

### ISSUE THREAD

**Issue by micheldenegri** [gnotaras/wordpress-wpvm#1]
Hello,
I've been having a problem when editing a page, in the log is showing this is the urls is trying to purge:
`2016-01-16T18:23:32+00:00 -- /about/([\d]+/)?$`

this is not actually purging anything, if i use the Purge single URL ins the backend, Im able to purge the url, and i can see this in the log:
`2016-01-16T18:26:44+00:00 -- /about/`

So i thing there is something wrong with that regexp, i get this error when i try that url in varnish:
```
varnish> ban.url '/about/([\d]+/)?$'
100        
Unknown request.
Type 'help' for more info.
Syntax Error: Invalid backslash sequence
```

any ideas what could be the problem?

**Comment by gnotaras** [comment]
Hi,

The problem most probably is the unescaped backslash of `\d`.

The `([\d]+/)?$` part of the regex is there to catch multipage content (posts/pages) and as far as I can tell the plugin sends the URL with the backslash escaped. Example from my varnishncsa log:

`
... [16/Jan/2016:20:04:15 +0000] "BAN http://www.g-loaded.eu/2013/02/13/back-to-blogging/([\\d]+/)?$ HTTP/1.0" 200 130 "-" ...
`

Apart from this, the `ban.url` command is really weird. I don't use it, but instead I use `ban` like this:

`
ban req.http.host == "www.g-loaded.eu" && req.url ~ "/2013/02/13/back-to-blogging/([\\d]+/)?$"
`

Please notice the escaped backslash.

Hope this helps.

**Comment by gnotaras** [comment]
> as far as I can tell the plugin sends the URL with the backslash escaped.

That's not true. Just checked varnishlog. The escaping is auto performed by varnish most probably. Opened a ticket in my private bug tracker to recheck this.

**Comment by gnotaras** [comment]
May I ask how do you perform the bans in `vcl_recv`? I highly recommend constructing a string like the following and supplying it to the `ban()` function:

```
req.http.host == g-loaded.eu && req.url ~ ^/2013/02/13/back-to-blogging/([\d]+/)?$
```

**Comment by micheldenegri** [comment]
I see, we currently dont have any ban method implemented on our `vcl_recv`, we are connecting to varnishadm port 6082 with our secret. Do you think this may be related? do you recommend instead to not use the varnishadm?

we currently have installed this version `"wpackagist-plugin/wpvm": "2.0.4"`

**Comment by gnotaras** [comment]
@micheldenegri 

Oh I see. This is the least tested functionality as I don't use it myself. It's there as a remnant from the original WP-Varnish (WPVM is a fork of WP-Varnish).

Most probably this mode of functionality has an issue with the non escaped backslash.

Here is something you could try until this is fixed:

1. Edit `https://github.com/gnotaras/wordpress-wpvm/blob/master/wpvm.php#L529`

The following:

```
$out = "ban req.url ~ ^$wpv_url$ && req.http.host == $wpv_host\n";
```

Should become:

```
$out = "ban req.url ~ ^" . addslashes($wpv_url) . "$ && req.http.host == $wpv_host\n";
```

2. In the WPVM settings select varnish version 3, check 'Use admin port" and make sure that the varnish secret is correct, and save the settings.

3. Check whether banning works as expected.

**Comment by gnotaras** [comment]
OK, I tested it and it seems to work for me without any changes. One thing I made a mistake with was to add the wrong varnish port in the WPVM settings. If the admin port it used, then port has to be set to 6082 or to whatever varnish's control port listens to.

**Comment by gnotaras** [comment]
Bug found. Line 529 of `wpvm.php` should be:

```
$out = "ban req.url ~ ^" . addslashes($wpv_url) . " && req.http.host == $wpv_host\r\n";
```

Changes:

1. Enclose `$wpv_url` in `addslashes()`
1. Remove `$` from `req.url` regex.
1. End with CRLF

Please try this corrected line if possible and let me if it works as expected.

**Comment by gnotaras** [comment]
>    Enclose $wpv_url in addslashes()
>    Remove $ from req.url regex.
>    End with CRLF

After making the aforementioned changes in line 529, purging using the control port works as expected in my tests. Looking forward to your feedback.

PS: I'm sorry about the goof ups. I was in the middle of doing something else and wasn't 100% focused on this.

**Comment by micheldenegri** [comment]
Excellent!, confirmed is now purging correctly, thanks a lot, any plans of pushing this fix soon?

**Comment by gnotaras** [comment]
Done! I just released 2.0.5 which contains the fix. Please let me know if everything works fine now.

Thank you for your feedback. Now that I took a look at the code again, I realized the plugin needs some work in some areas (minor improvements, code clean up, etc). I'll try to do this as soon as I find some spare time.

I'll now close this issue.

Best Regards,
George

PS: I usually neglect this, but if you like this plugin, a small review at any time on wordpress org is always much appreciated!

**Comment by gnotaras** [comment]
Michel, thanks for your review at wordpress org. It is much appreciated!
